R/teal_data-datanames.R

```diff
@@ -43,10 +43,25 @@ setGeneric("datanames<-", function(x, value) standardGeneric("datanames<-"))
 setMethod("datanames<-", signature = c("teal_data", "character"), definition = function(x, value) {
   checkmate::assert_subset(value, names(x@env))
   x@datanames <- value
+  x@datanames <- sort_datanames(x@datanames, x@join_keys)
   methods::validObject(x)
   x
 })
 setMethod("datanames<-", signature = c("qenv.error", "character"), definition = function(x, value) {
   methods::validObject(x)
   x
 })
+
+
+#' @keywords internal
+sort_datanames <- function(datanames, joinkeys) {
+
+  child_parent <- sapply(
+    datanames,
+    function(name) teal.data::parent(joinkeys, name),
+    USE.NAMES = TRUE,
+    simplify = FALSE
+  )
+
+  union(unlist(topological_sort(child_parent)), datanames)
```

Comment:
Does performing `intersect` has any consequence to `datanames`?

With this suggestion, `datanames` always remains the same

Reply:
I think this `union(intersect(A, B), B)` is just `B`

```r
is_B <- function(A, B) {
  unionAB <- union(intersect(A, B), B)
  cat('union', paste(unionAB, collapse = ','), '\n')
  cat('B',     paste(B,       collapse = ','), '\n')
  
  identical(
    sort(union(intersect(A, B), B)),
    sort(B)
  )
}

is_B(letters[sample(1:10, size = 5)], letters[sample(1:10, size = 3)])

table(
  replicate(
    100,
    is_B(letters[sample(1:10, size = 5)], letters[sample(1:10, size = 3)])
  )
)
> TRUE 
>  100
```

Reply:
@averissimo I don't think so, this can be translated into

```r

# sorted but always a subset of datanames
# this is needed because join_keys can possibly contain different datanames
# as join_keys are not so stricltly protected
# - also insersect defines order based on the first argument
#   intersect(c("c", "b"), c("a", "b", "c"))
#   [1] "c" "b"
relational_datanames <- intersect(unlist(topological_sort(child_parent)), datanames)
union(relational_datanames, datanames)
```

Reply:
Ah, the set is the same, but the order is different

Reply:
> I think this `union(intersect(A, B), B)` is just `B`

@m7pr you can't set `datanames` based on names from `join_keys` as `join_keys` might contain unknown datanames:

```r

data <- teal_data() |> within(iris <- iris)
join_keys(data) <- join_keys(join_key("a", "b", key = "id"))
datanames(data) <- "iris"
join_keys(data)
# A join_keys object containing foreign keys between 2 datasets:
# a: [no primary keys]
#   <-- b: [id]
# b: [no primary keys]
#   --> a: [id] 
datanames(data)
# [1] "iris"
```

Reply:
Ok, After yesterday's discussion I was under the assumption that we wanted to modify the `datanames` and extend with parent datasets.

If we're keeping the same `datanames` then all good!

Reply:
But in this case `child_parent` is only `iris` so I don't think we need to call `intersect` inside `union`.
I think we either call just `intersect(unlist(topological_sort(child_parent)), datanames)` or just `union(unlist(topological_sort(child_parent)), datanames)` - the latter is better as brings names of parent for limited `datanames()` that do not contain parent name.

```r
> datanames
[1] "iris"
> joinkeys
A join_keys object containing foreign keys between 2 datasets:
a: [no primary keys]
  <-- b: [id]
b: [no primary keys]
  --> a: [id] 
> child_parent
$iris
NULL

> topological_sort(child_parent)
[[1]]
[1] "iris"
```

Reply:
Okey @averissimo @m7pr  - I missed that `child_parent` object is already intersected. Dropping suggestion

```r
child_parent <- sapply(
    datanames,
    function(name) parent(joinkeys, name),
    USE.NAMES = TRUE,
    simplify = FALSE
  )
```

Reply:
@averissimo @m7pr 
Sorry for above comment. I was right, intersect is needed.

```r
data <- teal_data() |> within(b <- data.frame())
join_keys(data) <- join_keys(join_key("a", "b", key = "id"))
datanames(data) <- "b"
datanames(data)
# "a" "b"
```

Reply:
Wow, alright! Totally see this now - will accept this change, and will write few more tests

Reply:
But with this change we don't get `a` appended as parent when `b` is set in `datanames()`

```r
> sort_datanames
function(datanames, joinkeys) {

  child_parent <- sapply(
    datanames,
    function(name) parent(joinkeys, name),
    USE.NAMES = TRUE,
    simplify = FALSE
  )

  union(
    intersect(unlist(topological_sort(child_parent)), datanames), # join_keys can be set for inexisting datanames
    datanames
  )
}
<environment: namespace:teal.data>
> data <- teal_data() |> within(b <- data.frame(), a <- data.frame())
> join_keys(data) <- join_keys(join_key("a", "b", key = "id"))
> datanames(data)
character(0)
> datanames(data) <- "b"
> datanames(data)
[1] "b"
```
I thought we were planning to extend with parent, so this should return `c("a", "b")`

Reply:
So that's a pitty. We use `join_keys` to extend `datanames()` with `parent` name, but we can't defend from names that do 
not appear in `teal_data` because we only rely on `datanames()` that can be set as custom.

Reply:
It means that your argument about setting @datanames in a `join_keys` setter about keeping `object@datanames` and `datanames(object)` consistent isn't relevant neither.

----

If `datanames()` can potentially return names of objects which doesn't exist in `@env` then when the error should occur? 
1. when setting datanames
2. when getting datanames
3. in teal when trying to pull inexisting dataset from @env

IMO error should be thrown at 1.

Reply:
> It means that your argument about setting @datanames in a join_keys setter about keeping object@datanames and datanames(object) consistent isn't relevant neither.

It doesn't matter if `sort_datanames()` is applied in `join_keys` setter or in `datanames` getter. If we allow `x@datanames` to be different than `datanames(x)` we still can end up with a parent not being added to `datanames` on getter if we go with

```r
  union(
    intersect(unlist(topological_sort(child_parent)), datanames), # join_keys can be set for inexisting datanames
    datanames
  )
```

Reply:
I'll make separate branch for where we have `union(intersect(` and apply `sort_datanames()` in getter of `datanames()` to compare

Reply:
I made a separate branch https://github.com/insightsengineering/teal.data/tree/datanames_getter%40topological_sort%40669_insertUI%40main so it's easier to switch 

- `sort_datanames()` uses `union(intersect(`
- `sort_datanames()` only applied on a getter of `datanames()`

And the result is the same - `intersect(` limited the set so the parent is not appended to `datanames()`

```
data <- teal_data() |> within(b <- data.frame(), a <- data.frame())
join_keys(data) <- join_keys(join_key("a", "b", key = "id"))
datanames(data) <- "b"
datanames(data)
> [1] "b"
```

Reply:
@m7pr result is not the same when one specifies unsorted datanames

```r
pkgload::load_all("teal.data")
data <- teal_data() |> within({
  a <- data.frame()
  b <- data.frame()
}) 
join_keys(data) <- join_keys(join_key("a", "b", key = "id")) 
datanames(data) <- c("b", "a")
datanames(data)
```

Which solves your issue in `teal`, where:
- you wanted to **sort** datanames topologically 
- but not add datasets based on parent-child relationship

Reply:
> If datanames() can potentially return names of objects which doesn't exist in @env then when the error should occur?

Yeah, let's extend the setter, so it checks for non-existing names in `ls(get_env(teal_data())` and then we can go with a regular `union(unlist(topological_sort(), datanames)` without the `interesct()` needed

Reply:
> but not add datasets based on parent-child relationship

But I thought we **WANT** to add parents. If we don't want to add parents, then `intersect(` is needed

Reply:
For this

<img width="325" alt="image" src="https://github.com/user-attachments/assets/74018d0a-aa1a-4bda-a26b-f0ae619df9b6">


I get `"a" "b"` which is what we wanted. Ordered by the `join_keys` order